### ISSUE TITLE
Completions extremely slow

### ISSUE THREAD

**Issue by impact27** [spyder-ide/spyder#15042]
### What steps will reproduce the problem?

The completions are extremely slow. To reproduce:
0 - Disable / don't install Kite
1 - Open a new file
2 - type:
```
import numpy as np
np.<tab>
```
3 - Wait 15 seconds
4 - The completion appears

## Versions

* Spyder version: 4.3.0.dev0 
* Python version: 3.8.6
* Qt version: 5.12.10
* PyQt5 version: 5.12.3
* Operating System: Darwin 20.3.0

### Dependencies

```

# Mandatory:
applaunchservices >=0.1.7       :  0.2.1 (OK)
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  3.0.4 (OK)
cloudpickle >=0.5.0             :  1.6.0 (OK)
diff_match_patch >=20181111     :  20200713 (OK)
intervaltree >=3.0.2            :  3.1.0 (OK)
IPython >=7.6.0                 :  7.18.1 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
keyring >=17.0.0                :  21.4.0 (OK)
nbconvert >=4.0                 :  6.0.7 (OK)
numpydoc >=0.6.0                :  1.1.0 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.7.2 (OK)
pygments >=2.0                  :  2.7.1 (OK)
pylint >=1.0                    :  2.6.0 (OK)
pyls >=0.36.2;<1.0.0            :  0.36.2 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.3.0             :  0.3.0 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  1.0.1 (OK)
qtconsole >=5.0.1               :  5.1.0.dev0 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
setuptools >=39.0.0             :  50.3.0 (OK)
sphinx >=0.6.6                  :  3.2.1 (OK)
spyder_kernels >=1.10.0;<1.11.0 :  1.11.0.dev0 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog >=0.10.3               :  0.10.3 (OK)
zmq >=17                        :  19.0.2 (OK)

# Optional:
cython >=0.21                   :  0.29.21 (OK)
matplotlib >=2.0.0              :  3.3.2 (OK)
numpy >=1.7                     :  1.19.2 (OK)
pandas >=1.1.1                  :  1.1.3 (OK)
scipy >=0.17.0                  :  1.5.2 (OK)
sympy >=0.7.3                   :  1.6.2 (OK)
```

**Comment by impact27** [comment]
I have tried profiling this, and the time is taken by the jedi `get_signatures` function.
I have tried changing in `pyls/plugins/jedi_completion.py`:
```
def _format_completion(d, include_params=True):
    completion = {
        'label': d.name,
        'kind': _TYPE_MAP.get(d.type),
        'detail': _detail(d),
        'documentation': _utils.format_docstring(d.docstring()),
        'sortText': _sort_text(d),
        'insertText': d.name
    }
    return completion
```
And the completion returns to reasonable times. I guess I am loosing on the parameters, but it is worth it to avoid waiting 15 seconds for completion.

**Comment by ccordoba12** [comment]
Hey @impact27, thanks for reporting. This is a problem we plan to tackle after we release Spyder 5.

**Comment by RSKothari** [comment]
@ccordoba12 I am currently using Spyder 5 and the simplest of typing operations such as copy, paste and code completion takes a very long time. Is a fix imminent?

**Comment by raphaelquast** [comment]
i am having the very same issue for quite some time now...  (e.g. simple typing, copy-paste & code-completion are extremely slow)
from a quick look it seems there are already several other issues that might be related... (#14792 #14752 #14744 #14384) 

so far i tried some workarounds but the problem seems to re-appear all the time: 
  - re-start / shut down python-language-server
  - disable kite
  - disable completion & linting
  - avoid showing the outline
  - avoid working on network-drives

@ccordoba12 can you think of any temporary fix for this?  
(I'd love to continue using Spider 5.0.1 but right now it slows down to a point where it is barely usable any more)

Let me know if I can provide assistance in tracking down this issue!

**Comment by ccordoba12** [comment]
There are some packages (e.g. Numpy and Pandas) for which completions are extremely slow. But that's something we can't fix in Spyder but in the Python language server.

However, for standard library modules (e.g. `math` or `subprocess`) it should work really fast (it does in my case). Could you check that?

> (I'd love to continue using Spider 5.0.1 but right now it slows down to a point where it is barely usable any more)

Does that mean that things are working as expected in Spyder 4.2.5?

**Comment by raphaelquast** [comment]
@ccordoba12 
thanks for the response! 
well... I'm heavily using both `numpy` and `pandas`, but it's not just completion that's slow... 
it seems all is somehow affected... e.g. typing, copying, saving files etc. even if I disable the language-server!

 I'll install 4.2.5 and report my findings!

**Comment by ccordoba12** [comment]
> it seems all is somehow affected... e.g. typing, copying, saving files etc. even if I disable the language-server!

Are you using macOS Big Sur and Anaconda (not Conda-forge)?

**Comment by raphaelquast** [comment]
no, I'm mostly on Windows 10 and Windows Server 2012 R2 and I always use conda-forge

**Comment by raphaelquast** [comment]
@ccordoba12 OK, I've tested with a fresh install of spyder 4.2.5 (in a fresh python 3.8 environment that includes `pandas`, `numpy` `scipy` and several other packages) For a couple of hours it seemed to work fine... however at one stage everything started to slow down again and now also a restart does not help for long.

there is no unusual cpu or ram usage (and there is plenty of ram available) 
the only thing that popped up in the internal console is this:
```WARNING:traitlets:No such comm: 6a1aa661ad1711ebb58bf02f7421e2f3```

**Comment by ccordoba12** [comment]
By disabling code snippets in

    Tools > Preferences > Completion and linting > General

completions will be almost instantaneous after the first time they are triggered for packages like Numpy, Pandas and Matplotlib.

This will be available since our 5.1.0 version, to be released at the end of the month.

**Comment by Nomia** [comment]
there is no `General` tab under `Completion and Linting` section  any more in Spyder version 4.2.5 @ccordoba12 

I've tried disable everything in under `Completion and Linting`, it's still sluggish when typing, aka the typing latency is big

I've issued a bug at #16067 
but no reply for now, I really tried  almost all the solution on the web, nothing works
could u give me any help?

**Comment by ccordoba12** [comment]
This fix will only be available in Spyder 5.1.0, as I clearly mentioned above.
